### PR TITLE
Multiplatform - Native

### PR DESCRIPTION
**Pull request by luca992**
Added my native implementation targeting all tier 1-3 native targets.

Personally I don't agree with having a `NATIVE_TARGETS_ENABLED` toggle. But up to you.

Not sure why detekt fails for `Pbkdf2Sha512` in the `nonJvmMain` due to `Pbkdf2Sha512.F`:
```
> Task :detektAll FAILED
/Users/lucaspinazzola/Projects/kotlin-bip39/bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt:54:17: Function names should match the pattern: [a-z][a-zA-Z0-9]* [FunctionNaming]
```

But succeeds in `commonMain`. I renamed both to `Pbkdf2Sha512.f` for consistency

Edit: `linuxArm64` target wasn't added due to [okio not supporting it.](https://github.com/square/okio/issues/1171)

### CODE REVIEW

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt
Comment (ccjernigan) [inline bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt:116]: Since we're re-implementing this, does it make sense to drop the nullable for `salt` so we don't need the null check here?

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt
Comment (ccjernigan) [inline bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt:36]: Is there a way to reduce using Okio for this?

**Review by ccjernigan — COMMENTED** [review]
Thanks for the effort that went into creating this PR.  The changes are clear, and it is nice to have various native targets.  I appreciate the care that went into generating secure random data in platform-specific ways.

I've looked at this from the general Kotlin/Gradle code review, clarity, maintainability, etc. perspective, and there are some questions inline.  Another member of our team will be looking at this from a security perspective, and will chime in over the next week or so.

From my perspective, the main thing that we need to figure out is how we can reduce any third party dependency (e.g. Okio).  It's great that we don't introduce dependencies in the Android targets, and I hope we can build on that.

I added a [followup issue](https://github.com/zcash/kotlin-bip39/issues/163) for CI jobs for other platforms, as I see that as complementary but out of scope for this initial PR.  Relatedly, we may want to explicitly disable native release publishing until we get the CI jobs properly set up (we'll leave the publishing on for snapshots).  That would mean adding 
`ORG_GRADLE_PROJECT_NATIVE_ENABLED: false`
[here](https://github.com/zcash/kotlin-bip39/blob/c0e0c17b629ed03b7ab92eaa11cf13e8dfae97e2/.github/workflows/deploy-release.yml#L75).

**Comment by luca992** [comment]
these are [okios implementations](https://github.com/square/okio/tree/cd31c079e6935f4b8035baa989acff5f0caa6c7f/okio/src/hashFunctions/kotlin/okio/internal). I guess they could be copied. I don't necessarily agree with the no dependencies rule you have as java's standard library is pretty massive (essentially the already bundle things that would be a dependency in like any other language). 

Plus rolling your own crypto is something I avoid; I guess if you have the team to maintain it that works. But at that point honestly I would make a new library or module and import it into this project. That way other people can at least benefit from the code if you plan on maintaining it. If every project made everything self contained it would lead to tons of code duplication.

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/bip39/ToSha256.kt
Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/bip39/ToSha256.kt:5]: https://github.com/zcash/kotlin-bip39/pull/162#issuecomment-1426613778

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt
Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt:70]: https://github.com/zcash/kotlin-bip39/pull/162#issuecomment-1426613778

File: bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt
Comment (hashben228) [inline bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt:15]: What errors occur if this @Suppress is removed? Are there consequences to making the value nullable instead of suppressing the unchecked cast?

File: bip39-lib/src/commonMain/kotlin/cash/z/ecc/android/bip39/Mnemonics.kt
Comment (luca992) [reply bip39-lib/src/commonMain/kotlin/cash/z/ecc/android/bip39/Mnemonics.kt:29]: that would break the actual purpose of it being a `var`:

```
   var cachedList = WordList()

    fun getCachedWords(languageCode: String): List<String> {
        if (cachedList.languageCode != languageCode) {
            cachedList = WordList(languageCode)
        }
        return cachedList.words
    }
```

... I think I removed `internal` by accident. I'll fix that. However it could also be made `private`. It's not used anywhere else.

Comment (luca992) [reply bip39-lib/src/commonMain/kotlin/cash/z/ecc/android/bip39/Mnemonics.kt:29]: also the suppress isn't technically needed. it will print a bunch of `info` level logs while building however:
`With old Native GC, variable in singleton without @ThreadLocal can't be changed after initialization`

Which is just annoying considering kotlin uses the new GC by default now

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt
Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt:137]: changed it to just:
```
        password?.fill(' ')
        password = null
        
```

Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt:137]: made a few changes to use `.isEmpty()` and `.isNotEmpty()` in that file that the ide suggested

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt
Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/PBEKeySpecCommon.kt]: yeah, good point

File: bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt
Comment (luca992) [reply bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt:15]:
```
> Task :bip39-lib:compileKotlinMingwX64 FAILED
e: warnings found and -Werror specified
w: file:///Users/lucaspinazzola/Projects/kotlin-bip39/bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt:17:33 Unchecked cast: CPointer<ByteVar /* = ByteVarOf<Byte> */> to CPointer<UByteVar /* = UByteVarOf<UByte> */>

Execution failed for task ':bip39-lib:compileKotlinMingwX64'.
> Compilation finished with errors
```

Comment (luca992) [reply bip39-lib/src/mingwMain/kotlin/cash/z/ecc/android/random/SecureRandom.kt:15]: it's supposed to be a UByte not a Byte for mingw. The suppress isn't for null-ablity

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt
Comment (luca992) [reply bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt]: removed its use

**Comment by luca992** [comment]
Turns out the native tests weren't running at all. Fixed that by applying the kotest plugin.

And that led me to find out there was an issue with the require statement in the native nextBytes implementation. Had to change the check: `check(result < 0)` -> `check(result >= 0)`

**Comment by luca992** [comment]
I think I've fixed everything that needs to be fixed. Up to you guys on how you want to handle `hmacSha512` and `sha256`

File: gradle/wrapper/gradle-wrapper.properties
Comment (ccjernigan) [inline gradle/wrapper/gradle-wrapper.properties:3]: We need to revert this Gradle version bump.  We are blocked on upgrading Gradle until all our other repos can also be upgraded.  This restriction is to support Gradle included builds between SDK, app, and bip-39, so we bump Gradle, Kotlin, and Android Gradle Plugin in lockstep between these projects to avoid compatibility issues.  All of this is currently blocked by incompatibility in the Rust Gradle Plugin with AGP 7.4.1+ in the SDK.

File: bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt
Comment (ccjernigan) [inline bip39-lib/src/nonJvmMain/kotlin/cash/z/ecc/android/crypto/Pbkdf2Sha512.kt:39]: Thanks for rolling back use of Okio here.  This minimizes the surface area of using third parties, and will make it easier to swap out Okio with something else in the future since its now just the hashing functions.

File: bip39-lib/src/commonMain/kotlin/cash/z/ecc/android/bip39/Mnemonics.kt
Comment (ccjernigan) [reply bip39-lib/src/commonMain/kotlin/cash/z/ecc/android/bip39/Mnemonics.kt]: Thanks—reviewing just on GitHub hides some of the important context!

**Review by ccjernigan — COMMENTED** [review]
Thanks for these enhancements.  I have a few minor comments inline, and I'm about to trigger the CI build.

File: gradle/wrapper/gradle-wrapper.properties
Comment (luca992) [reply gradle/wrapper/gradle-wrapper.properties]: reverted to 7.6

**Comment by luca992** [comment]
found another option for the hashing:
https://github.com/korlibs/korge/tree/main/krypto/src

**Review by hashben228 — APPROVED** [review]

**Comment by ccjernigan** [comment]
@luca992 Thanks for your patience.  We are still having internal security auditing discussions, and we hope to wrap those up soon.  As a preview, this where I think we will land:
 - Merge this PR as-is (I might ask to tweak the readme slightly because it's explicit about no dependencies which doesn't hold true with this PR).  This PR moves our multiplatform efforts forward in a positive way in the codebase, while native deployments are still initially disabled.
 - Create a followup issue to swap out Okio.  We're leaning towards copying the necessary hashing functions from Okio.  Then we can own and audit those as our own implementation going forward.
 - Enable deployment of native targets on CI

No action items yet, just keeping you in the loop as we're making decisions internally.

**Comment by ccjernigan** [comment]
@luca992 Although it took a while, I got internal approval for the plan discussed above https://github.com/zcash/kotlin-bip39/pull/162#issuecomment-1450062550

Once we resolve the conflicts, then I can merge this PR 🎉

**Comment by luca992** [comment]
@ccjernigan I updated kotlin and all dependencies. If you don't want that lmk and I'll revert it

**Comment by ccjernigan** [comment]
> @ccjernigan I updated kotlin and all dependencies. If you don't want that lmk and I'll revert it

We don't mind updating dependencies, although I would recommend a separate PR for the dependencies. When I merge this PR, it will be squashed into a single commit.  For dependencies, Kotlin 1.8.21 is good to go, and we've updated our other repos as well.  (We're sticking with Gradle 7.6 until we resolve an issue with the Rust integration in another project though).

We still have merge conflicts with this PR—I think the WordList changes from the main branch need to be brought in which incorporates a significant performance improvement.

**Review by hashben228 — APPROVED** [review]

**Review by ccjernigan — APPROVED** [review]
Thank you for the effort that went into this PR, as well as your patience during our review process.  This brings some great enhancements for the project and the ecosystem.

**Comment by ccjernigan** [comment]
I created a new PR that incorporated these changes while resolving the remaining conflict.  These changes have been merged. See https://github.com/zcash/kotlin-bip39/pull/188

**Comment by luca992** [comment]
Np. Now I'll make that last pr for JS 👌

**Pull request closed without merging**
